### ISSUE TITLE
Lagging while using nano, vim, etc.

### ISSUE THREAD

**Issue by svaderia** [jwilm/alacritty#2217]
Which operating system does the issue occur on?
`Ubuntu 16.04`
If on linux, are you using X11 or Wayland?
`X11`

When I use nano it isn't loaded immediately, after I press a key it gets rendered. even while typing in nano, A key press will load the previous key and not the current key. Even when exiting, the save file option is shown only after I press some key. This issue remains with any editor I use. Even when using `git log --graph --pretty='\''%Cred%h%Creset -%C(yellow)%d%Creset %s %Cgreen(%cr) %C(bold blue)<%an>%Creset'\'' --abbrev-commit` command, it gets rendered only after I press some key. Following is the video to showcase the issue. 

![simplescreenrecorder-2019-03-26_14 22 30](https://user-images.githubusercontent.com/26005890/54984228-6ce32f80-4fd4-11e9-98c0-f0c079789787.gif)

If you need any other info related to system config or want me to try out some experiments, I can help.

EDIT 1 : Using `alacritty 0.2.7`

**Comment by svaderia** [comment]
Another thing I noticed is, everything works perfectly when I use a new instance of alacritty. But when an instance stays inactive for a while, it starts lagging. Sometimes writing in the terminal starts lagging.

**Comment by cole-h** [comment]
Can you reproduce this in another terminal emulator? What if you try a different shell (`bash`, `zsh`, `fish`)? And finally, can you see if this is reproducible on the latest master? I don't have this issue and that's what I'm currently running. Even checking out the 0.2.7 tag doesn't give me the same behavior.

**Comment by chrisduerr** [comment]
Could you try running Alacritty with `alacritty --print-events` to show if Alacritty is updated with a delay or if the key events are actually coming in *after* the key is pressed (with a delay).

**Comment by svaderia** [comment]
@cole-h 

> What if you try a different shell (`bash`, `zsh`, `fish`)?

The issue remains irrespective of what I use.


> Can you reproduce this in another terminal emulator?

I'm unable to reproduce.


> can you see if this is reproducible on the latest master?

I will try it and let you know.

**Comment by svaderia** [comment]
@chrisduerr, Following is the output of the `print-events`. 

```
[2019-03-26 23:57] [INFO] Welcome to Alacritty
[2019-03-26 23:57] [INFO] Configuration loaded from "/home/shyamal/.config/alacritty/alacritty.yml"
[2019-03-26 23:57] [INFO] Device pixel ratio: 1.0833333333333333
[2019-03-26 23:57] [INFO] Width: 1366, Height: 744
[2019-03-26 23:57] [INFO] Initializing glyph cache...
[2019-03-26 23:57] [INFO] ... finished initializing glyph cache in 0.033508283s
[2019-03-26 23:57] [INFO] Width: 1366, Height: 744
[2019-03-26 23:57] [INFO] Cell Size: (10 x 21)
[2019-03-26 23:57] [INFO] Padding: (0 x 0)
[2019-03-26 23:57] [INFO] PTY Dimensions: Line(35) x Column(136)
[2019-03-26 23:57] [INFO] Initialisation complete
[2019-03-26 23:57] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: Resized(LogicalSize { width: 1023.6923076923077, height: 686.7692307692308 }) }
[2019-03-26 23:57] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: Moved(LogicalPosition { x: 0.0, y: 22.153846153846157 }) }
[2019-03-26 23:57] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: Resized(LogicalSize { width: 1260.923076923077, height: 686.7692307692308 }) }
[2019-03-26 23:57] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: Refresh }
[2019-03-26 23:57] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: Focused(true) }
[2019-03-26 23:57] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: CursorMoved { device_id: DeviceId(X(DeviceId(2))), position: LogicalPosition { x: 0.0, y: -22.153846153846157 }, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
[2019-03-26 23:57] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 28, state: Released, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:57] [INFO] Font size changed to Size(24) with DPR of 1.0833333333333333
[2019-03-26 23:57] [INFO] Width: 1366, Height: 744
[2019-03-26 23:57] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 28, state: Released, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:57] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 49, state: Pressed, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 49, state: Pressed, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('n') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 49, state: Released, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 49, state: Released, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 30, state: Pressed, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 30, state: Pressed, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('a') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 30, state: Released, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 30, state: Released, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 49, state: Pressed, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 49, state: Pressed, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('n') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 49, state: Released, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 49, state: Released, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 24, state: Pressed, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 24, state: Pressed, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('o') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 24, state: Released, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 24, state: Released, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 24, state: Pressed, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 24, state: Pressed, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('o') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 24, state: Released, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 24, state: Released, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 14, state: Pressed, virtual_keycode: Some(Back), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 14, state: Pressed, virtual_keycode: Some(Back), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('\u{8}') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 14, state: Released, virtual_keycode: Some(Back), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 14, state: Released, virtual_keycode: Some(Back), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 106, state: Pressed, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 106, state: Pressed, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 106, state: Released, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 106, state: Released, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 28, state: Pressed, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 28, state: Pressed, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('\r') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 28, state: Released, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 28, state: Released, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 106, state: Pressed, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 106, state: Pressed, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 106, state: Released, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 106, state: Released, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 30, state: Pressed, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 30, state: Pressed, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('a') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 30, state: Released, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 30, state: Released, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 31, state: Pressed, virtual_keycode: Some(S), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 31, state: Pressed, virtual_keycode: Some(S), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('s') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 31, state: Released, virtual_keycode: Some(S), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 31, state: Released, virtual_keycode: Some(S), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter(' ') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 32, state: Pressed, virtual_keycode: Some(D), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 32, state: Pressed, virtual_keycode: Some(D), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('d') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 32, state: Released, virtual_keycode: Some(D), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 32, state: Released, virtual_keycode: Some(D), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter(' ') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 33, state: Pressed, virtual_keycode: Some(F), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 33, state: Pressed, virtual_keycode: Some(F), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('f') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 33, state: Released, virtual_keycode: Some(F), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 33, state: Released, virtual_keycode: Some(F), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter(' ') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter(' ') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter(' ') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter(' ') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 48, state: Pressed, virtual_keycode: Some(B), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 48, state: Pressed, virtual_keycode: Some(B), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('b') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 48, state: Released, virtual_keycode: Some(B), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 48, state: Released, virtual_keycode: Some(B), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Pressed, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter(' ') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 57, state: Released, virtual_keycode: Some(Space), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 29, state: Pressed, virtual_keycode: Some(LControl), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 29, state: Pressed, virtual_keycode: Some(LControl), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 45, state: Pressed, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 45, state: Pressed, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: true, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('\u{18}') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 29, state: Released, virtual_keycode: Some(LControl), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 29, state: Released, virtual_keycode: Some(LControl), modifiers: ModifiersState { shift: false, ctrl: true, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 45, state: Released, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 45, state: Released, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 106, state: Pressed, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 106, state: Pressed, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 106, state: Released, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 106, state: Released, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 49, state: Pressed, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 49, state: Pressed, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('n') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 49, state: Released, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 49, state: Released, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 18, state: Pressed, virtual_keycode: Some(E), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 18, state: Pressed, virtual_keycode: Some(E), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('e') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 18, state: Released, virtual_keycode: Some(E), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 18, state: Released, virtual_keycode: Some(E), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 45, state: Pressed, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 45, state: Pressed, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('x') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 45, state: Released, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 45, state: Released, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 23, state: Pressed, virtual_keycode: Some(I), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 23, state: Pressed, virtual_keycode: Some(I), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('i') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 23, state: Released, virtual_keycode: Some(I), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 23, state: Released, virtual_keycode: Some(I), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 20, state: Pressed, virtual_keycode: Some(T), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 20, state: Pressed, virtual_keycode: Some(T), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('t') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 20, state: Released, virtual_keycode: Some(T), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 20, state: Released, virtual_keycode: Some(T), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 28, state: Pressed, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 28, state: Pressed, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: ReceivedCharacter('\r') }
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: Awakened
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 28, state: Released, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] glutin event: WindowEvent { window_id: WindowId(X(WindowId(69206021))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 28, state: Released, virtual_keycode: Some(Return), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
[2019-03-26 23:58] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(13))), event: Key(KeyboardInput { scancode: 106, state: Pressed, virtual_keycode: Some(Right), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } }) }
[2019-03-26 23:58] [INFO] Goodbye

```

**Comment by chrisduerr** [comment]
@svaderia I'm sorry if I've been a bit unspecific, but the event log itself doesn't actually help. I'd like to know if the events arrive when you press the keys, or if they are delayed.

The goal here is to find out if the key events arrive late, or if they arrive as soon as you press the key, but Alacritty doesn't update for a while.

**Comment by svaderia** [comment]
@chrisduerr I got your point. I checked it, the event arrives at the time of key press but alacritty doesn't update till I press the next key. 

Another thing I noticed is, the delay only happens when I'm using alacritty in full screen mode (That is my default and preffered choice). It works perfectly fine when I resize it.

**Comment by chrisduerr** [comment]
> @chrisduerr I got your point. I checked it, the event arrives at the time of key press but alacritty doesn't update till I press the next key.

That's very odd, it would mean that Alacritty is not redrawing appropriately. Do you use compton as a compositor or a full package like Gnome/KDE?

**Comment by svaderia** [comment]
I'm using Gnome and not Compton.

**Comment by chrisduerr** [comment]
If you have the chance to easily test on Wayland, it might be interesting to see if that changes things.

Some hardware information might also be useful. Which GPU and driver are you using?

**Comment by svaderia** [comment]
> If you have the chance to easily test on Wayland, it might be interesting to see if that changes things.

Will let you know.

> Some hardware information might also be useful. Which GPU and driver are you using?

GPU : GM108M [GeForce 840M]
driver : i915

**Comment by chrisduerr** [comment]
>  driver : i915

That's an intel iGPU driver though, right? So probably not what's being used for the 840M? If you're actually using the Nvidia GPU with the correct driver, that shouldn't be an issue though.

**Comment by svaderia** [comment]
Ah, my bad. Using `NVIDIA binary driver - version 387.26 from Nvidia-387`

**Comment by pappasam** [comment]
@chrisduerr I'm using Alacritty's latest "master" (91aa683bcd060b2ac2f621a388a6448f564d0537) and am experiencing this issue with any program that makes frequent updates to the terminal (eg, apt install <pkg>,  vim [:PlugUpdate], etc). Few screen updates take place until I press a key on my keyboard, then everything updates at once.

This issue does NOT exist in Alacritty v0.2.9 for me; everything works flawlessly there.

Both "latest master" and "v0.2.9" were compiled from source on my machine.

OS: Linux Mint 19.1
Laptop: Dell XPS 15
GPU: Nvidia Driver Version 390.116-0
       description: 3D controller
       product: GP107M [GeForce GTX 1050 Mobile]
       vendor: NVIDIA Corporation
       physical id: 0
       bus info: pci@0000:01:00.0
       version: a1
       width: 64 bits
       clock: 33MHz
       capabilities: pm msi pciexpress bus_master cap_list rom
       configuration: driver=nvidia latency=0
       resources: irq:136 memory:ec000000-ecffffff memory:c0000000-cfffffff memory:d0000000-d1ffffff ioport:e000(size=128) memory:ed000000-ed07ffff

**Comment by pappasam** [comment]
See below for a screencast using Vim and the latest version of Alacritty. I'm using a program to print my keys at the end so you can see when they're pressed:
![screencast](https://user-images.githubusercontent.com/3723671/55291365-dcc63100-53ab-11e9-964b-32804fb6b8ce.gif)

**Comment by laanwj** [comment]
> When I use nano it isn't loaded immediately, after I press a key it gets rendered. even while typing in nano,

FWIW I'm not sure if it's the same issue, but I recently get the issue that the terminal, while running an interactive process (such as a `wget` download) *stops updating*. It doesn't hang, it continues fine, but it stops updating the screen. Only after a keypress the new state of the terminal appears.

**Comment by laanwj** [comment]
I bisected it to:
```
0f96a62218cfc78b56606f0dcc334d59202edf7c is the first bad commit
commit 0f96a62218cfc78b56606f0dcc334d59202edf7c
Author: Christian Duerr <chrisduerr@users.noreply.github.com>
Date:   Thu Mar 14 20:06:38 2019 +0000

    Update glutin to version 0.20
    
    Fixes #458.
    Fixes #1681.

:100644 100644 8c8cb0a95d517b5385ed6fac6a77a09d9186245a 67535e3f0bd7636858d1f9f738151712c0bad7e8 M	CHANGELOG.md
:100644 100644 3370ad4ea01f69a7a55f9400a8c94e1770df827e 619deec9662a4d48abd801351c705d45f1967459 M	Cargo.lock
:100644 100644 b6a7d975b29131507c5859d1b2f9b2158e88ccc4 f5818866e25430794230703ea4401640deda55b0 M	Cargo.toml
:040000 040000 79a4a644c582f0c4b7e089a7cf0d8a96bede0209 e6ea040c63a3233ef2968ad58ebe2c864f3fd1b4 M	src
```

**Comment by chrisduerr** [comment]
It's surprising to me that this would be caused by a glutin update. Just to make sure, can you verify this with the latest master? I've recently bumped glutin to 0.21 so maybe that fixed it.

Otherwise it seems like we'd have to bisect this in winit which would likely be a very laborious task especially because I can't do it myself since I can't reproduce this.

**Comment by laanwj** [comment]
I'm surprised too, but I'm fairly sure about it, I tried again: 
- I get the behavior with 0f96a62218cfc78b56606f0dcc334d59202edf7c 
- and not with its parent commit b1032bcc6b79135f87f327548e43563da05657fb
- also have the problem with master (as of now, f0dbfc9b9c6514bc60e271bb74c2970e645d998f)

Hopefully others can confirm this.

> Otherwise it seems like we'd have to bisect this in winit which would likely be a very laborious task especially because I can't do it myself since I can't reproduce this.

Why winit? I'd say the next step would be bisecting glutin between 0.20 and 0.19?

**Comment by chrisduerr** [comment]
Glutin is mostly just a wrapper around winit. Most of the changes are actually made in winit and not glutin. So I'd expect this breakage to come from a winit update and not a glutin update.

The reason why this is caused by a glutin update is likely just because the winit version was bumped in glutin 0.20 to the next major winit version and we interact with winit exclusively through glutin.

**Comment by chrisduerr** [comment]
The only glutin-related issue I can see would be if the `Awakened` event isn't propagated properly. For that reason it might be interested to look at where exactly the freeze is happening in Alacritty.

If anyone is willing to help with troubleshooting, putting a println before and after [this statement](https://github.com/jwilm/alacritty/blob/master/src/event.rs#L482-L485) could give some insights. I'd assume it's stuck in there and it would be really interesting if that's not the case.

**Comment by chrisduerr** [comment]
Oh, I've found something that might be extremely relevant and which is simple to test. In the `src/window.rs` we have a `with_vsync(true)` option. If anyone could try to disable that it would be very helpful.

According to [winit's documentation](https://docs.rs/winit/0.19.0/winit/struct.EventsLoop.html#danger) this can cause some problems.

**Comment by laanwj** [comment]
> If anyone is willing to help with troubleshooting, putting a println before and after this statement could give some insights. I'd assume it's stuck in there and it would be really interesting if that's not the case.

I did this—and yes, it hangs before the statement. It also seems *any kind* of event will wake it up, not just a keypress, but also mouse movement.

**Comment by chrisduerr** [comment]
And I'd assume while hanging, no `Awakened` events are displayed in `--print-events`?

Can you check if the wakeup is attempted by adding some debug output [here](https://github.com/jwilm/alacritty/blob/f0dbfc9b9c6514bc60e271bb74c2970e645d998f/src/window.rs#L452)? That would clearly indicate what's going on.

I'm assuming that Alacritty is trying to wake up the event loop (which is sleeping), but the wakeup call never arrives so no `Awakened` event is printed in `--print-events` and the loop is never woken up.

**Comment by laanwj** [comment]
> Can you check if the wakeup is attempted by adding some debug output here? That would clearly indicate what's going on.

Done. Below (the terminal in the background on the top-right shows the debug info) is output with the debug lines added and `--print-events`. Sorry for the bad photo, I can't make a screenshot as any touching of the computer causes it to become unstuck:

![screen](https://user-images.githubusercontent.com/126646/55683598-cc7cfb80-5941-11e9-82af-c9674101a676.jpg)

So yes I'd say it looks like the wakeup is attempted but it stays in wait_events.

**Comment by chrisduerr** [comment]
Reported upstream in https://github.com/tomaka/winit/issues/834. Thanks for the help with troubleshooting.

If anyone is willing to go through the effort of bisecting, please report your results there. If you need help with it, please let me know.

**Comment by svaderia** [comment]
Any workarounds till this get fixed?

**Comment by pappasam** [comment]
@chrisduerr I've done some sleuthing and have a path for future research: https://github.com/tomaka/winit/issues/834#issuecomment-482682007

If anyone on this thread is familiar with X Events, specifically, with the following man pages, you may find the explanation more quickly than me:

* https://linux.die.net/man/3/xcheckifevent
* https://linux.die.net/man/3/xnextevent

**Comment by chrisduerr** [comment]
Thanks a ton for looking into this with so much detail. Hopefully that will help upstream.

**Comment by pappasam** [comment]
@chrisduerr the problem goes away when I remove the explicit "wait_for_event" in src/event.rs:

![image](https://user-images.githubusercontent.com/3723671/56070349-646d5180-5d55-11e9-9e74-a64964d37f08.png)

"window.wait_events" ultimately calls winit's "run_forever" function, which relies on select (2).

@chrisduerr can you evaluate whether and why the "wait_for_event" flag is still useful / necessary? It only lives in src/event.rc and its relevant commit mentions something about "visual bell support". This was its date/author/summary message:

2017-02-03|Mark Andrus Roberts|Add visual bell support

At first glance, by removing this step, the terminal performance seems to get a lot faster and this issue becomes resolved.

Unless you feel strongly otherwise, I'll create a PR for you to evaluate.

**Comment by chrisduerr** [comment]
> can you evaluate whether and why the "wait_for_event" flag is still useful / necessary?

It is absolutely necessary, yes. Everything else would just be wasting resources completely, it's entirely unrelated to the visual bell.

**Comment by pappasam** [comment]
Based on observation, this issue only occurs with multithreaded programs. For me, [vim-plug](https://github.com/junegunn/vim-plug).

It only hangs during the threaded calls. If I force the program to use 1 thread with :PlugUpdate1, there is no problem.

[line showing vim-plug is multithreaded](https://github.com/junegunn/vim-plug/blob/master/plug.vim#L1733)

I've done a lot of digging but have come up short. A lot of things changed in winit, glutin AND alacritty in early March, making debugging this issue incredibly challenging.

**Comment by chrisduerr** [comment]
I don't think `nano` is heavily multithreaded and in general I don't see how this would be affected by multithreading. So I think that might be a red herring or another issue (like a vim-internal one).

**Comment by chrboe** [comment]
I'm having the same issue, and I think it's worth noting that (for me) it only occurs when using tmux. I can reliably reproduce this problem by running a command like `htop -d 1` and waiting a few seconds; alacritty then stops drawing until I press a key. This does not happen when using alacritty without tmux.

**Comment by laerus** [comment]
i can reproduce this reliably with everything that streams output in Alacritty, pressing a random key (even without Alacritty having focus) resumes the stream printing,

i can reproduce this with either Nvidia or Intel graphics cards and it doesn't happen on other terminals

**Comment by traviscross** [comment]
This bash loop is sufficient to reproduce the issue:

```sh
for ((i=0;; i++)); do printf '\r%s' $i; sleep .1; done
```

Eventually the count will appear to hang.  Keyboard input (even shift or alt-tab) will cause it to jump up to the correct value and resume counting for awhile.

On current master (f3cc7c4c4e46778eedab18c4d25c7009b9bbb04f), this reproduces both with and without tmux.  In general, it appears to reproduce somewhat more quickly and reliably under tmux.

Confirming the bisection results of others above, we found that commit 0f96a62218cfc78b56606f0dcc334d59202edf7c ("Update glutin to version 0.20") reproduces the issue but the commit before it (b1032bcc6b79135f87f327548e43563da05657fb "Add text reflow") does not.

**Comment by chrisduerr** [comment]
The issue can be reproduced and has been identified. It is waiting only on upstream progress.

There's no further information necessary, so please keep things to a minimum until the upstream issue has progressed.

**Comment by zachliu** [comment]
Same issue as @pappasam when using `0.3.2`

**Comment by zachliu** [comment]
> I'm having the same issue, and I think it's worth noting that (for me) it only occurs when using tmux. I can reliably reproduce this problem by running a command like `htop -d 1` and waiting a few seconds; alacritty then stops drawing until I press a key. This does not happen when using alacritty without tmux.

I cannot reproduce it using `tmux` (`tmux 2.9a`) and `htop -d 1`. 

OS: Linux Mint 19.1
Laptop: Dell XPS 13
Kernel: x86_64 Linux 4.15.0-51-generic
Shell: zsh 5.4.2

**Comment by Syndicate6690** [comment]
Same issue here, whenever I run something (does not depend on which shell, terminal multiplexer, or application) that runs for more than a few seconds of outputting text to the tty, it just freezes until my mouse is moved or I press a key. 

OS: Arch Linux amd64, Kernel 5.1.7, zsh 5.4.1, i3-gaps
Device: HP Elitebook 9470M, i7-3687U, 8GB RAM

**Comment by jstkdng** [comment]
Can confirm it happens to me, though, only when using the open source drivers. Like amdgpu, nouveau, i915, etc. It doesn't happen with the nvidia propietary driver.

**Comment by SealedJoy** [comment]
+1 same issue
arch linux - x11 - zsh
latest alacritty from arch repo 0.3.3-3
amd - opencl-mesa 19.1.4-1

**Comment by laanwj** [comment]
I've tried, and haven't seen the issue anymore after #2438, while it was consistently happening beforehand. Thanks for fixing this!

**Comment by leosunmo** [comment]
Would love a release with this fix in it. It's quite a big deal!

**Comment by NathanCrama** [comment]
+1. Is there a way to build it from source while waiting for a release with this fix?

**Comment by sagis-tikal** [comment]
@NathanCrama Just clone the repository, change branches if needed, using cargo build Alacritty (run `cargo build --release`) and copy the binary from target/release/ to your $PATH instead of the current binary.

**Comment by NathanCrama** [comment]
@sagis-tikal thanks for your answer. I guess my question wasn't clear enough: what I was wondering was which branch, if any, has this fix implemented?

**Comment by sagis-tikal** [comment]
@NathanCrama I used the master branch and it seems like the change is still there.

**Comment by nixpulvis** [comment]
FWIW @NathanCrama, with `git` and GitHub you can answer these questions yourself. GitHub shows which PR this issue was closed in above:

![DeepinScreenshot_select-area_20191111101640](https://user-images.githubusercontent.com/715947/68598236-6d99da80-046c-11ea-8485-e8b564d8d9d1.png)

Going to that PR we see that it was merged in 729eef0:

![DeepinScreenshot_select-area_20191111101800](https://user-images.githubusercontent.com/715947/68598327-9326e400-046c-11ea-9cf9-7734099202ff.png)

Now with an updated version of the alacritty repo locally, you can run `git log` and look for that commit to be sure it's in the tree. Or even run `git log 729eef0^..HEAD` to see all commits (including the merge) which came afterwards.

Hope this helps.

**Comment by chuckmasucci** [comment]
I've been using Master ever since the fix went in on 10/4 and it's working fine.

**Comment by NathanCrama** [comment]
I don't know, I'm really confused right now. I've experienced the bug yesterday on `alacritty-git` from the AUR on Manjaro but I'm unable to reproduce it today.

**Comment by nixpulvis** [comment]
Yea I was having trouble reliably triggering this issue too. Spooky behavior indeed. Luckily it happened reliably enough over the course of a day, we shouldn't have a problem confirming in practice that it's fixed.

**Comment by nixpulvis** [comment]
@NathanCrama I too couldn't reliably reproduce this issue, though I would see it from time to time. Once we have a few remaining issues resolved on mater we'll be cutting a new release.

Until then, you can either run master (or something with the merged PR) or wait until the release. Sorry for any inconveniences this has caused.

**Comment by Zebradil** [comment]
@NathanCrama I think you can reliably check from which commit your alacritty is built by looking at version at version number in the output of `pacman -Qi alacritty-git`.

> @NathanCrama Just clone the repository, change branches if needed, using Cargo, build Alacritty (run `cargo build --release`) and copy the binary from target/release/ to your $PATH instead of the current binary.

This is a good way for someone who doesn't want to install `alacritty-git` or doesn't have it available (non-AUR distro). One note to that, though: I'd recommend changing PATH instead of replacing binary manually:
```bash
export PATH="${HOME}/.cargo/bin:${PATH}"
```
With this approach it's enough to do `cargo install --path .` in one's working copy of alacritty.
After required version is released `cargo uninstall alacritty` can be used to remove locally built version and use the one from distro repos.

**Comment by jeffhiltz** [comment]
I'm also using `alacritty-git` from the Arch User Repository (AUR) and I'm still having the problem, or one like it.

I have been using `alacritty 0.3.3 (3475e449)` which was built Oct 11, 6 days after #2438  was merged to master.

I've just upgraded to `alacritty 0.3.3 (561063b5)` which is from today but don't expect to see any difference.

**Comment by kodaman2** [comment]
Same issue on lubuntu 19.04, vim typing lags by a few milliseconds. What you type, is seen on the screen a few milliseconds later. If you type very fast, you can def see the huge lag to the point where is not bearable. Full screen is lagging more than minimized to a smaller size for sure, but something is def bogging down performance.

I don't see this issue with terminator or qterminal.

Perhaps a good Troubleshoooting.md could help users to get to the bottom of this.

**Comment by jeffhiltz** [comment]
I realized that I didn't update this thread after fixing my problem.  Sorry about that!

I changed my compton backend from `glx` to `xr_glx_hybrid` and poof!  Delay is gone.  The issue wasn't with Alacritty (for me.)

**Comment by kodaman2** [comment]
@jeffhiltz I'll give it a try since I'm on lubuntu is a good chance it could fix my issue. Cheers!
